### PR TITLE
feat: support extmark_overlay

### PR DESCRIPTION
**Pull request by windwp**
It allow highlight character different with current character on buffer
it working with hop and lightspeed.

lightspeed need  #869 to have same result like neovim

### CODE REVIEW

**Comment by windwp** [comment]
@andyljones @theol0403

**Comment by andyljones** [comment]
Fantastic! I pulled #868 and #869 into the same branch and ran the extension, and it all seems to work! Happy to pay out soon as it's merged.

**Comment by theol0403** [comment]
Great to see this implemented! Funny timing as I was just starting to have the time to give it a shot, but it seems like you've done a great job of the PR :). I'll have to test this more thoroughly, but I've noticed some bugs:
- viewport height broken: I know you've mentioned this, but lightspeed scans much too far out out of the viewport, affecting performance, but also does not work when you scoll down beyond 200 lines. Hopefully the only solution is not to patch lightspeed, and instead provide a way for the native nvim api to know the vscode viewport.
- "instant mode" cursor, which comes after sneaking or f/t-ing, the fake cursor is hidden. Solution is to do `hi LightspeedCursor gui=reverse`
- eol labels: lightspeed supports doing `s<CR>` which then shows a label *after* the last character. These are currently hidden.
- I'll have to figure out how to reproduce, but I've found that sometimes the 1-character highlighting produces some weird overlay text

**Comment by windwp** [comment]
about the viewport height issue. I think @asvetliakov  already try to sync visible range to vscode but it have problem. 
I already try and the scrolling is not good. I am not sure about jumplist.

 https://github.com/vscode-neovim/vscode-neovim/blob/master/src/cursor_manager.ts#L583
 https://github.com/vscode-neovim/vscode-neovim/blob/master/vim/vscode-neovim.vim#L89
 that why i create #869  and do a patch to lighspeed.
 https://gist.github.com/windwp/5fd492d8ace55fbe12bb7be67594bb60
It is a hacking way but it working well and don't break anything.

> I'll have to figure out how to reproduce, but I've found that sometimes the 1-character highlighting produces some weird 
> overlay text

if that character has the same character on buffer then it won't use overlay method. It highlight same as a normal character but with different color. It makes better performance with another plugin because some plugin highlight ext_mark for background text too so every character is overlay and it slow vscode.

- I will fix eol issue and add test.

**Comment by justinmk** [comment]
> I think @asvetliakov already try to sync visible range to vscode but it have problem. I already try and the scrolling is not good. I am not sure about jumplist.
> 
> https://github.com/vscode-neovim/vscode-neovim/blob/master/src/cursor_manager.ts#L583 https://github.com/vscode-neovim/vscode-neovim/blob/master/vim/vscode-neovim.vim#L89

That's an old hack using winline(). Since then, Nvim RPC UI has added the `win_viewport` UI event: 

https://github.com/neovim/neovim/blob/f876a0eeb5af6f41acb187eb9442062c3984b26d/runtime/doc/ui.txt#L603

**Comment by theol0403** [comment]
> That's an old hack using winline(). Since then, Nvim RPC UI has added the `win_viewport` UI event:
> 
> [neovim/neovim@`f876a0e`/runtime/doc/ui.txt#L603](https://github.com/neovim/neovim/blob/f876a0eeb5af6f41acb187eb9442062c3984b26d/runtime/doc/ui.txt#L603)

This should also fix #149. @windwp it would be awesome if you could implement this!

**Comment by windwp** [comment]
I think vscode neovim already implement win_viewport.
https://github.com/vscode-neovim/vscode-neovim/blob/b9def15f267d0e212c989981c883741de4256993/src/highlight_manager.ts#L88-L99
https://github.com/vscode-neovim/vscode-neovim/blob/52ff62612d3bb9ae2a50ce00b9e068c5d175abda/src/cursor_manager.ts#L109

vscode know the viewport of neovim but that viewport is not same of current visible ranges on vscode.

neovim viewport height is 200 and cursor always on a center position . `set scrolloff=100`.
https://github.com/vscode-neovim/vscode-neovim/blob/master/vim/vscode-options.vim#L16

**vscode manage all scroll function** . Neovim doesn't have API to scroll and spam <c-e> <c-d> is not good.

https://github.com/vscode-neovim/vscode-neovim/blob/master/vim/vscode-neovim.vim#L89

**Comment by andyljones** [comment]
> viewport height broken: I know you've mentioned this, but lightspeed scans much too far out out of the viewport, affecting performance, but also does not work when you scoll down beyond 200 lines. Hopefully the only solution is not to patch lightspeed, and instead provide a way for the native nvim api to know the vscode viewport.

Impressed that you spotted this - I was trying to get a minimal example of this behaviour earlier and was pretty puzzled.

**Comment by trkoch** [comment]
> Fantastic! I pulled https://github.com/vscode-neovim/vscode-neovim/pull/868 and https://github.com/vscode-neovim/vscode-neovim/pull/869 into the same branch and ran the extension, and it all seems to work!

I gave this a try with `vim-sneak` and label mode enabled. The overlay itself seems to work now. Great! I noticed, however, the cursor is no longer placed on the first occurrence (as it is still is with label mode disabled). I need to e.g. press `<space>` to do so (or press `w` which moves from the found word onwards). Also, the first occurrence is highlighted in a less visible way than following occurrences. I know this is very early in progress, but wanted to share my findings with you.

**Comment by windwp** [comment]
@trkoch  you should try to change highlight  color on vim-sneak label.

**Comment by theol0403** [comment]
Rebase and tests should work now

**Comment by Shatur** [comment]
@theol0403 Is there anything else blocking the merge?

File: package.json
Comment (justinmk) [inline package.json:1306]: can this avoided? sinon is pretty jank. And we've avoided it up to now.

Comment (windwp) [reply package.json]: ok

**Review by justinmk — APPROVED** [review]
nice! thanks for getting rid of sinon. Dependencies can really become a maintenance liability.

**Comment by theol0403** [comment]
> @theol0403 Is there anything else blocking the merge?

Sorry for the delay. There are two things:

1. Avoidance of lightspeed patch. Ideally, this PR would include a solution that does not involve patching lightspeed. This is so that all plugins do not need to be patched like hop/sneak/leap. I was hoping @justinmk could provide a suggestion to https://github.com/vscode-neovim/vscode-neovim/pull/868#issuecomment-1111619159, which basically says that there exists an api for `nvim scrolling->vscode scrolling` but not `vscode scrolling -> nvim set scrolling`. It would be ideal if @windwp could contribute a solution that allows nvim and vscode scrolling to be synchronized, fixing jump plugins and also C-u/C-d scrolling.
2. Some bugs I haven't had time to report:
		- highlighting tabs cause heart symbol to appear as overlay: 
image
		- also, searching in lightspeed in a buffer with tabs cause hearts
		- vscode previous cursor does not disappear when lightspeed tries to hide it (minor)
		- once, I experienced that the grey wash did not disappear after lightspeed was done, and I had to make a new search for the grey to go away (minor)

If the strange heart issue is resolved, I can merge this.

**Comment by windwp** [comment]
@theol0403  can you provide a step to get "heart char"

**Comment by windwp** [comment]
it is the solution for scrolling,
https://github.com/vscode-neovim/vscode-neovim/blob/1442396107c7314af024e6fdd8832cc06dc0792e/src/cursor_manager.ts#L583
https://github.com/vscode-neovim/vscode-neovim/blob/master/vim/vscode-neovim.vim#L89
maybe it have another issue so @asvetliakov comment it.

 lightspeed doesn't need scrolling to work it only need to know the viewport of vscode because on the first key `s` it find a unique text  on screen, another jump target plugin doesn't have that feature.

so  i suggest to make a new option to allow scroll neovim by <c-e> or <c-y> . we can uncomment that code and see what happen.

**Comment by theol0403** [comment]
> @theol0403 can you provide a step to get "heart char"

Copy this C file with tabs:

```c
static uint8_t notify_func(struct bt_conn *conn,
			   struct bt_gatt_subscribe_params *params)
{
	if (!data) 
		printk("[UNSUBSCRIBED]\n");
	printk("[NOTIFICATION] data %p length %u\n", data, length);
	return BT_GATT_ITER_CONTINUE;
}
```

Line select all to get:

image

My theme: Monokai Pro

**Comment by theol0403** [comment]
Thanks for the fix! I found one more bug:
* eol on empty line: lightspeed shows labels when doing `s<CR>` on empty lines. However, label isn't shown on those empty lines in vscode.

I'll merge after this, we can look into the scrolling issues separately. I realized that I've been incorrectly assuming what the problem is, its less major than I thought. I had thought that lightspeed didn't work below line 200 at all, not that it just doesn't know about scrolling (but does know about position). 

I'd say the other significant problem is that lightspeed tries to match all 200 lines in the buffer, which reduces performance and accuracy of lightspeed. The solution is to reduce the nvim buffer size to the vscode window size and solve issues related to that - PRs welcome!

**Comment by theol0403** [comment]
Looks like extmark tests fail on latest commit. Will merge once fixed.

**Comment by theol0403** [comment]
Thank you for the PR! And thanks @andyljones for the incentive :)

Just a reminder that for optimum lightspeed UX:
- `hi LightspeedCursor gui=reverse` for fake cursor
- https://github.com/vscode-neovim/vscode-neovim/pull/869 and https://gist.github.com/windwp/5fd492d8ace55fbe12bb7be67594bb60 for proper viewport height for labeling 

@windwp your code is very well written and you have a great understanding of nvim and this plugin. Even after this PR, would you consider being a contributor? Thanks!

**Comment by windwp** [comment]
@theol0403  Thank i will try to make a sync viewport on neovim currently it have issue with fold.

Maybe we can rewrite some of vim code to lua.  Create _G.vscode and vscode.api.

**Comment by unTanana** [comment]
Hey, great work here :D One question though. Is there any way to disable this feature / highlighting? thank you.

**Comment by windwp** [comment]
@theol0403  is this your email theol0403@gmail.com.
@unTanana it  only highlight when plugin call it

**Comment by theol0403** [comment]
> @theol0403 is this your email [theol0403@gmail.com](mailto:theol0403@gmail.com). 
Yep! Thanks!

**Comment by unTanana** [comment]
Hmmmm, this might be a stupid think to ask, is this the highlight in question?
image

Because it seems to be:
1. Activated when using visual mode for selecting multiple lines (which is fine I guess)
2. Not going away after I'm done selecting
3. Not going away after I'm done jumping with an easymotion plugin either.

If the issue I'm describing is unrelated to this change, I apologise

**Comment by theol0403** [comment]
> Hmmmm, this might be a stupid think to ask, is this the highlight in question? image

Actually, I've noticed that too, but I didn't realize it was from this PR. It appears as though tabs are highlighted extra. Not sure why they aren't going away though.

**Comment by unTanana** [comment]
I can confirm that downgrading fixes this, and also my issue with easymotion excessively triggering the output tab, covering my screen each time I want to do a jump

**Comment by Muzzamil1** [comment]
Hi, thanks for adding support for the lightspeed plug-in.

1. There is a delay using the "f" and "F" keys.
2. With "s" and "S", the cursor won't jump for the first match (white highlighted match) until I press escape, which is kind of confusing sometimes as it is not sure whether we jumped to that place or not until press escape

**Comment by kevinhwang91** [comment]
Sorry for my laziness to open a new issue. I post the issue I am encountering, looks like has some bugs for some plugins using virttext with overlay. 
Using https://github.com/kevinhwang91/nvim-fFHighlight, 
image

**Comment by theol0403** [comment]
> Hi, thanks for adding support for the lightspeed plug-in.
> 
> 1. There is a delay using the "f" and "F" keys.
> 2. With "s" and "S", the cursor won't jump for the first match (white highlighted match) until I press escape, which is kind of confusing sometimes as it is not sure whether we jumped to that place or not until press escape 

https://github.com/vscode-neovim/vscode-neovim/pull/868#issuecomment-1131963354

**Comment by asvetliakov** [comment]
Just in case this code is very old (from the first versions) and i'm not sure if it would work anymore.

**Comment by theol0403** [comment]
> @theol0403 Thank i will try to make a sync viewport on neovim currently it have issue with fold.
> 
> Maybe we can rewrite some of vim code to lua. Create _G.vscode and vscode.api.

Sorry, I missed this. Awesome, thanks a lot! And yes those sound like great places to start!

**Pull request merged**
